### PR TITLE
feat: migrate to typescript

### PR DESCRIPTION
**Pull request by matmalkowski**
This is initial Pull Request with the goal of having brand new, simplified API & native TS support. The change contains few main areas:
 - I've used the code from @partheseas fork as suggested, with some clean up in the structure 
- I've re-did the build system to fit TS package management - using rollup to build the bundle 
- Repo structure has been changed - we don't need monorepo for the 2 projects here, and it would only mean pain in the release process, considering the e2e package was private and without dependencies
- I've started with some basic e2e test of the current implementation: simple main & render thread stores synced together and the counter component that updates visuals on both render & main
- Added typescript types compatibility test agains basic redux usage. We also should consider extending it with applyMiddleware and other usages that might be breaking.
- And ofc GH action to build & test it all so far.

Target branch of this PR is `v2` - follow up PRs would setup the `semantic-release` to help with the deployment process and the `v2` branch would be considered an `alpha` version.

The code itself from my look at it is still rough on the edges and needs some work, I've even added some `TODO:` comments in the most obvious places, but overall I believe the goal should to be to polish it on the `v2` branch with incremental refactors/fixes/tests.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,38 +1,102 @@
...
-  "version": "1.5.1",
-  "description": "Using redux in electron",
-  "main": "index.js",
-  "repository": "https://github.com/klarna/electron-redux",
-  "author": "Burkhard Reffeling <burkhard.reffeling@klarna.com>",
+  "version": "2.0.0-alpha.0",
+  "description": "Redux & Electron: Make sure all your stores are on the same page",
+  "repository": "https://github.com/klarna/electron-redux.git",
+  "authors": [
+    "Burkhard Reffeling <burkhard.reffeling@klarna.com> (https://github.com/gaearon)",
```
Comment (hardchor) [inline package.json:7]: Not sure if @gaearon appreciates, but that's not my github name 😅

File: src/syncMain.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+	stopForwarding,
+	validateAction,
+} from "./utils";
+
+const middleware: Middleware = (store) => {
+	ipcMain.handle("electron-redux.FETCH_STATE", async () => {
+		// Stringify the current state, and freeze it to preserve certain types
+		// that you might want to use in your state, but aren't JSON serializable
+		// by default.
+		return JSON.stringify(store.getState(), freeze);
```
Comment (matmalkowski) [inline src/syncMain.ts:22]: Actually I don't like idea of having this by default baked into the lib - it goes kinda against the redux recommendations itself:
> It is highly recommended that you only put plain serializable objects, arrays, and primitives into your store.
I would like to extract this and make it configurable so users when using the middleware can configure it with their own serializer at their own risk (and we ofc can provide them with example serializer used here 👍🏻 )

File: src/syncRenderer.ts [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+		return next(action);
+	};
+};
+
+export const syncRenderer: StoreEnhancer = (createStore: StoreCreator) => {
+	preventDoubleInitialization();
+
+	return (reducer, state) => {
+		const store = createStore(
+			wrapReducer(reducer as any), // TODO: this needs some ❤️
```
Comment (matmalkowski) [inline src/syncRenderer.ts:85]: This cast needs some work like mentioned in the comment

File: src/syncRenderer.ts [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+		);
+
+		// This is the reason we need to be an enhancer, rather than a middleware.
+		// We use this (along with the wrapReducer function above) to dispatch an
+		// action that initializes the store without needing to fetch it synchronously.
+		getRendererState((state) => {
+			store.dispatch(replaceState(state));
+		});
+
+		// TODO: this needs some ❤️
```
Comment (matmalkowski) [inline src/syncRenderer.ts:97]: And this magic TS fix is another thing that needs to get looked at

File: src/utils/actions.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
+import { FluxStandardAction, isFSA } from "flux-standard-action";
+
+// Certain actions that we should never replay across stores
+const blacklist = [/^@@/, /^redux-form/];
```
Comment (matmalkowski) [inline src/utils/actions.ts:4]: also seems something that we might consider configurable in the future (but maybe not now)

**Review by sameoldmadness — APPROVED** [review]
LGTM 👍

**Review by victorhqc — APPROVED** [review]
Thanks for the PR, the code looks good to me. There are things that probably need more discussion about namings probably, and documentation, specially a migration guide if we're planning to introduce breaking changes in API.

Tried to run in a project but the current implementation doesn't work with middlewares, so this could probably be worked on in the next PR.

**Pull request merged**